File: AlgoAnalysis-C/CountingSort/CountingSort.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>
#include<math.h>
#include<limits.h>

#define ARRAYSIZE 100000
#define COUNTING 10000

void COUNTING_SORT(int A[], int B[], int k);

int main()
{
    int numbers[ARRAYSIZE];
    int sorted_numbers[ARRAYSIZE];
    int i,j;
    int num;
    time_t t;
    srand((unsigned) time(&t));
    for(i = 0; i < ARRAYSIZE; i++)
    {
        num = rand()%COUNTING;
        numbers[i] = num;
    }

/*    for(i = 0; i < ARRAYSIZE; i++)
        printf("%d: %d\n", i+1, numbers[i]);
*/
    clock_t start, end;
    double cpu_time_used;

    start = clock();
    COUNTING_SORT(numbers,sorted_numbers,COUNTING);
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("Time taken by Counting Sort: %f sec\n", cpu_time_used);
/*    for(i = 0; i < ARRAYSIZE; i++)
        printf("%d: %d\n", i+1, sorted_numbers[i]);
    return 0;
*/
}
void COUNTING_SORT(int A[], int B[], int k){

    // 1. ADIM: SAYAÇ DİZİSİNİ OLUŞTURMA VE SIFIRLAMA
    // Sayıların frekansını (kaç tane olduğunu) tutacak 'c' dizisi oluşturuluyor.
    int c[k];

    // c dizisinin içinde rastgele (çöp) değerler kalmasın diye
    // baştan sona tüm elemanları 0'a eşitliyoruz.
    for (int i = 0; i < k;i++) {
        c[i] = 0;
    }

    // 2. ADIM: FREKANSLARI SAYMA (COUNTING)
    // Asıl dizimiz olan A dizisini baştan sona geziyoruz.
    for (int j = 0; j < ARRAYSIZE; j++) {
        // A[j] değerini bir indis olarak kullanıyoruz.
        // Örneğin sayı 5 ise, c dizisinin 5. indisindeki değeri 1 artırıyoruz.
        // Böylece o sayıdan kaç tane olduğunu kaydetmiş oluyoruz.
        c[A[j]] = c[A[j]] + 1;
    }

    // 3. ADIM: KÜMÜLATİF TOPLAM (POZİSYON BELİRLEME)
    // Burada c dizisini güncelliyoruz. Her elemanı kendinden öncekiyle topluyoruz.
    // Bu işlem, sayıların sıralı dizide (B dizisinde) hangi son indiste
    // yer alacağını belirlememizi sağlar.
    for (int i = 1; i < k;i++) {
        c[i] = c[i] + c[i - 1];
    }

    // 4. ADIM: SIRALI DİZİYİ (B) OLUŞTURMA
    // A dizisini SONDAN başa doğru tarıyoruz (Algoritmanın kararlı olması için).
    for (int j = ARRAYSIZE - 1; j >= 0;j--) {

        // A[j]'deki sayıyı alıyoruz, c dizisinden bu sayının nereye konulacağına bakıyoruz.
        // İndisler 0'dan başladığı için 1 çıkarıp B dizisindeki yerine koyuyoruz.
        B[c[A[j]]- 1] = A[j];

        // Aynı sayıdan bir tane daha gelirse, bir önceki pozisyona (sola)
        // yerleşmesi için sayacın değerini 1 azaltıyoruz.
        c[A[j]] = c[A[j]] - 1;
    }

}

File: AlgoAnalysis-Python/CountingSort/CountingSort.py
import random
import time
from array import array

ARRAYSIZE = 1000
COUNTING = 1000


def counting_sort(A, B, k):
    # 1. ADIM: SAYAÇ DİZİSİNİ OLUŞTURMA (INITIALIZATION)
    # 0'dan k-1'e kadar olan sayıların frekansını tutacak C listesini oluşturuyoruz.
    # Python'da '[0] * k' diyerek içi 0 dolu, k boyutunda bir liste tek satırda açılır.
    # (C'deki döngüyle 0'a eşitleme işleminin Pythoncası budur).
    C = [0] * k

    # 2. ADIM: FREKANSLARI SAYMA (COUNTING)
    # Giriş dizisi A'yı baştan sona tarıyoruz.
    for j in range(0, ARRAYSIZE):
        # A[j] değerini indeks olarak kullanıp, C dizisindeki karşılığını 1 artırıyoruz.
        # Yani "bu sayıdan bir tane daha buldum" diyerek sayacı güncelliyoruz.
        C[A[j]] = C[A[j]] + 1

    # 3. ADIM: KÜMÜLATİF TOPLAM (POZİSYON BELİRLEME)
    # C dizisindeki her elemanı kendinden öncekiyle topluyoruz.
    # Bu işlem sonucunda C[i], i sayısının sıralı B dizisinde yerleşeceği
    # "en son indeksi" (limit pozisyonunu) belirtir hale gelir.
    for i in range(1, k):
        C[i] = C[i] + C[i - 1]

    # 4. ADIM: SIRALI DİZİYİ (B) OLUŞTURMA (PLACING)
    # A dizisini SONDAN başa doğru tarıyoruz.
    # range(başlangıç, bitiş-hariç, adım) -> (ARRAYSIZE-1'den başla, -1'e kadar git, 1 azalt)
    # Sondan başlamamızın sebebi algoritmanın "kararlı" (stable) olmasını sağlamaktır.
    for j in range(ARRAYSIZE - 1, -1, -1):
        # Mantık zinciri:
        # 1. A[j]'yi al (Sıralanacak sayıyı seç).
        # 2. C[A[j]] değerine bak (Bu sayı B dizisinde hangi pozisyonda olmalı?).
        # 3. İndeksler 0'dan başladığı için 1 çıkar ve B dizisine yerleştir.
        B[C[A[j]] - 1] = A[j]

        # Aynı sayıdan bir daha gelirse (döngü devam ederken),
        # bir solundaki boşluğa yerleşsin diye pozisyon bilgisini 1 azaltıyoruz.
        C[A[j]] = C[A[j]] - 1


if __name__ == "__main__":
    numbers = [0] * ARRAYSIZE
    sorted_numbers = [0] * ARRAYSIZE

    for i in range(ARRAYSIZE):
        numbers[i] = random.randint(0, COUNTING - 1)

    for i in range(ARRAYSIZE):
        print(f"{i+1}: {numbers[i]}")

    start = time.time()
    counting_sort(numbers, sorted_numbers, COUNTING)
    end = time.time()

    cpu_time_used = end - start
    print(f"Time taken by Counting Sort: {cpu_time_used} sec")

    for i in range(ARRAYSIZE):
        print(f"{i+1}: {sorted_numbers[i]}")